### PR TITLE
Return response even for JSON encoding errors

### LINKED ISSUE #60: JSONEncoding errors don't return a response
If `registry.json_encoder` throws an exception, the exception is not returned as an error JSON response to the client.

### PR DESCRIPTION
**Pull request by Nifdee**
Fixes #60.

### CODE REVIEW

File: tests/test_registry.py [hunk 1]
```diff
@@ -645,6 +645,58 @@ def raise_exception():
...
+        response = json.loads(registry.dispatch(fake_batch_request))
+
+        assert isinstance(response, list)
+        assert len(response) == 2
+
+        foo_response = response[0] if response[1]["id"] == "bar" else response[1]
+        TestDispatch.assert_error(foo_response, "foo", InternalError)
+        assert foo_response["error"]["data"] == "Could not parse the input data."
+
+        bar_response = response[1] if response[1]["id"] == "bar" else response[0]
```
Comment (mnazbro) [inline tests/test_registry.py]: Can you put the response checking together:

    if response[0]["id"] == "foo":
        foo_response = response[0]
        bar_response = response[1]
   else:
        ...

File: tests/test_registry.py [hunk 2]
```diff
@@ -645,6 +645,58 @@ def raise_exception():
...
+        assert "error" not in bar_response
+
+    def test_encoder_exception_single(self):
+        def throw_error(input):
+            raise InternalError("Could not parse the input data.")
+
+        registry = Registry()
+        registry.json_encoder.encode = throw_error
```
Comment (mnazbro) [inline tests/test_registry.py]: This test should ruin all of the other tests (unless there's some magic reloading). You should use `@pytest.patch` to replace the encode function for the scope of this test.

File: tests/test_registry.py [hunk 3]
```diff
@@ -645,6 +645,58 @@ def raise_exception():
         assert "error" in response
         assert random_val == response["error"]["data"]["random"]
 
+    def test_encoder_exception_batch(self):
+        def custom_encode(input):
+            if "foo" in str(input):
+                raise InternalError("Could not parse the input data.")
+            else:
+                return json.JSONEncoder().encode(input)
```
Comment (mnazbro) [inline tests/test_registry.py]: I think this test would be more robust if you stored the original encoder and used it for the else case.

File: typedjsonrpc/registry.py [hunk 4]
```diff
@@ -126,6 +129,25 @@ def _handle_exceptions(self, method, is_notification=False, msg_id=None):
...
+    def _encode_complete_result(self, result):
+        if isinstance(result, list):
+            return '[' + ','.join([self._encode_single_result(res) for res in result]) + ']'
+        else:
+            return self._encode_single_result(result)
+
+    def _encode_single_result(self, result):
+        msg_id = Registry._get_id_if_known(result)
+        is_notification = msg_id is None
+        encoded = self._handle_exceptions(method=self.json_encoder.encode,
```
Comment (mnazbro) [inline typedjsonrpc/registry.py]: `method=` is unnecessary

File: typedjsonrpc/registry.py [hunk 5]
```diff
@@ -107,9 +107,12 @@ def _wrapped():
 
         return self._handle_exceptions(_wrapped, is_notification, self._get_id_if_known(msg))
 
-    def _handle_exceptions(self, method, is_notification=False, msg_id=None):
+    def _handle_exceptions(self, method, is_notification=False, msg_id=None, params=None):
```
Comment (mnazbro) [inline typedjsonrpc/registry.py]: I have a preference to keep this method simpler and have the method which is passed in capture the parameters inside it. That way we don't have to have an if-else statement in here which isn't really relevant to the exception handling code

File: typedjsonrpc/registry.py [hunk 6]
```diff
@@ -126,6 +129,25 @@ def _handle_exceptions(self, method, is_notification=False, msg_id=None):
...
+            return self._encode_single_result(result)
+
+    def _encode_single_result(self, result):
+        msg_id = Registry._get_id_if_known(result)
+        is_notification = msg_id is None
+        encoded = self._handle_exceptions(self.json_encoder.encode,
+                                          is_notification=is_notification,
+                                          msg_id=msg_id,
+                                          params=result)
+        if isinstance(encoded, dict):
```
Comment (mnazbro) [inline typedjsonrpc/registry.py]: I think this code would make more sense if you made `_handle_exceptions` return a tuple with whether or not there was an error.

That way we wouldn't have to type check!

File: tests/test_registry.py [hunk 7]
```diff
@@ -645,6 +646,66 @@ def raise_exception():
         assert "error" in response
         assert random_val == response["error"]["data"]["random"]
 
+    def test_encoder_exception_batch(self):
+        registry = Registry()
+        original_encode = registry.json_encoder.encode
```
Comment (mnazbro) [inline tests/test_registry.py]: You can implement this with a scoped `pytest.patch`. If you want some advice on actually how to do this, feel free to ask.

Implementing it yourself doesn't handle the case where an exception is thrown in the middle.

File: tests/test_registry.py [hunk 8]
```diff
@@ -645,6 +646,66 @@ def raise_exception():
...
+
+        def custom_encode(input):
+            if "foo" in str(input):
+                raise InternalError("Could not parse the input data.")
+            else:
+                return original_encode(input)
+
+        registry.json_encoder.encode = custom_encode
```
Comment (mnazbro) [inline tests/test_registry.py]: Delete this line

**Comment by mnazbro** [comment]
Looks good. @csivaguru Can you test this?

**Comment by csivaguru** [comment]
QE looks good

**Pull request merged**